### DISCUSSION TITLE
How to hold objects with lifecycle inside State?

### DISCUSSION THREAD

**Discussion by okmaybenot** [pointfreeco/swift-composable-architecture#3205] [Q&A] [answered]
Context:
I'm using Kotlin multiplatform and trying to share view models between iOS and Android.
I would also like to use TCA to manage navigation and UI on the iOS side.

Problem:
When using Observable objects, I know exactly when the view will be deallocated because the Observable object will call deinit and I can synchronize it to free the Kotlin view model (this requires a special ceremony that I do using a property wrapper).

Is it possible to get something similar with TCA State? Or maybe there is something else that will help solve my problem without manually managing the lifecycle (which is not always possible)

Example:
I open screen A and start listening a socket inside the view model. When I close screen A, I want to release the viewModel (and close the socket)

**Comment by okmaybenot** [comment, marked as answer]
Somewhat solved this using this trick, not sure it's the best way tho

```
case .onTask:

                return .run { @MainActor send in
                    
                    viewModel.create()
                    
                    await withTaskGroup(of: Void.self) { group in

                        group.addTask { @MainActor in
                            do { try await Task.never() } catch {
                                viewModel.clear()
                            }
                        }
```

**Comment by mbrandonw** [comment]
Hi @romanatexn, yes this is essentially the way to do it. The lifecycle of a TCA feature can be modeled by a long living effect since when TCA features are dismissed their effects are torn down (assuming you are using the presentation tools of the library, e.g. `ifLet`).

However, there are some things you should consider to improve this. First, if you are sending the `onTask` action from the `.task` or `.onAppear` view modifier, then you should know that is called every time the view appears. So, if you present something over the view and then dismiss it, the `onAppear`/`task` will call again.

So really you should introduce a kind of `onFirstAppear`. It does not need to be async, and you can find examples of this online if you don't want to write it yourself. And then you can write your code simply as:

```swift
case .onFirstAppear:
  return .run { @MainActor _ in 
    viewModel.create()
    try? await Task.never()
    viewModel.clear()
  }
```

**Reply by okmaybenot** [reply]
Thank you!
Fantastic improvements!
